[C#] Fix and improve project files

ConanChenTookMyName opened the pull request:
Fixes #6103.

Adds multitargeting to .NET Core project, and renames it to "FlatBuffers.csproj", to identify it as main project.

* "net46" target included for Unity 2017<sup>[[1]](https://docs.microsoft.com/en-us/visualstudio/cross-platform/unity-scripting-upgrade?view=vs-2019)</sup>
* "netstandard2.0" target included for Unity 2018<sup>[[2]](https://blogs.unity3d.com/2018/03/28/updated-scripting-runtime-in-unity-2018-1-what-does-the-future-hold/)</sup> and general use<sup>[[3]](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting)</sup>
* "netstandard2.1" target included for Span&lt;T&gt; support

Renames .NET Framework project to "FlatBuffers.net35.csproj". Legacy project is needed because no .NET Framework Developer Pack is available for targeting .NET Framework 3.5 using a .NET Core project and .NET Core SDK.

* "net35" target included for Unity 5<sup>[[1]](https://docs.microsoft.com/en-us/visualstudio/cross-platform/unity-scripting-upgrade?view=vs-2019)</sup>

Also adds project properties to the .NET Core project, for defining `UNSAFE_BYTEBUFFER`, `BYTEBUFFER_NO_BOUNDS_CHECK`, and `ENABLE_SPAN_T` conditional compiler constants using command line arguments at build time.

e.g.:
> `dotnet build -f "netstandard2.1" -p:UNSAFE_BYTEBUFFER=true -p:ENABLE_SPAN_T=true "FlatBuffers.csproj"`

aardappel commented:
Thanks! Is there anything that should be added to https://github.com/google/flatbuffers/blob/master/docs/source/CsharpUsage.md to guide people using this? Such as maybe your notes about these constants?

ConanChenTookMyName commented:
I originally found descriptions for these constants in "[net/FlatBuffers/ByteBuffer.cs](https://github.com/google/flatbuffers/blob/c75ae242931821a5e50c1221d425a2629adb871b/net/FlatBuffers/ByteBuffer.cs#L17-L35)". It seems they've been in the code for some time, but there just wasn't a convenient way to enable them (would've required using awkward syntax at the command line when building, or modifying the project file). I can copy the descriptions found in the source file to the C# documentation.

Should I also add a "Building the FlatBuffers C# library" section near the top of the documentation, and mention Unity support, since building FlatBuffers for C# seems to have a lot of additional considerations?

harujoh commented:
Hello.
I made a pull request earlier here, and it seems to be covered by this change.

https://github.com/google/flatbuffers/pull/6073

However, I made a mistake and was unable to make a pull request.

aardappel commented:
@ConanChenTookMyName ah, didn't know.. but yes, any hints you can add to the docs for future users will be helpful.

ConanChenTookMyName commented:
I added a "Building the FlatBuffers C# library" section to the documentation to explain the supported .NET and Unity versions, as well as a "Conditional compiler constants" section with the descriptions from ["net/FlatBuffers/ByteBuffer.cs"](https://github.com/google/flatbuffers/blob/c75ae242931821a5e50c1221d425a2629adb871b/net/FlatBuffers/ByteBuffer.cs#L17-L35). I also corrected a few minor inconsistencies in formatting and wording (e.g.: changed "C-sharp" to "C#").

aardappel commented:
Awesome, thanks for adding those.

harujoh commented:
Excuse me.

Why was this pull request merged first, knowing that it would obviously conflict?
Give me a good explanation.

Has @dbaileychess  gone on an extended vacation or something?

https://github.com/google/flatbuffers/pull/6073

aardappel commented:
Sorry, was not aware of the possible conflict.
@dbaileychess do we need to revert this one, or just deal with the conflicts in the other?

harujoh commented:
I think what we need is a policy decision.

@dbaileychess said "we want to keep flatbuffers as widely available as possible".
What is "widely available as possible".

I prefer to keep projects like this commit in one place.

However, I'm wondering why this merge of commits has taken place without the same discussion that took place in my pull request.

Why did you decide in the past that it was a good idea to create a separate pull request for the project files, separate from each other?

Please tell us about it.

ConanChenTookMyName commented:
The conflict is only in the "FlatBuffers.Core.csproj" project file. #6073's changes to the project file just adds "netstandard2.1" framework target, so their changes can be safely replaced with mine.

However, #6073 also removes the need for defining `UNSAFE_BYTEBUFFER` when using `ENABLE_SPAN_T`, so I would need to update my additions to the C# documentation to accommodate that.

It might be easiest to revert my changes for now, and I'll re-submit after #6073 is merged?

harujoh commented:
He's right, I don't need FlatBuffers.Core.csproj myself, and I have no objection to it being removed.

But that's not the end of the matter.

If FlatBuffers.Core.csproj is not needed, then my committed FlatBuffers.Core.Test.csproj should be aligned with FlatBuffers.Core.csproj and should not be created.

The reason I say there needs to be policy and discussion is because of these side conflicts.

aardappel commented:
Yes, fully agree with @dbaileychess, what that means is simply that as many users of C# can use it with all sorts of versions. Especially users of Unity.

dbaileychess commented:
Lets revert this, i'll prioritize getting #6073 merged, and then we can replay this (and fix any conflicts). 

@harujoh Sorry for not being responsive on #6073, flatbuffers is just a side project for me and I don't always have time to look at immediately. We don't want your work to be in vain, and appreciate the PR to make this project better.

dbaileychess commented:
Added revert at #6130 @aardappel if you agree with the plan you can submit the revert.

ConanChenTookMyName commented:
Don't worry about replaying this after #6073 is merged, as I'll need to make a few edits to the documentation included in this pull request to account for the `ENABLE_SPAN_T` usage changes in #6073. I'll submit a new pull request with the additional changes after.

dbaileychess commented:
@ConanChenTookMyName The revert and #6073 are merged, so go ahead with your updates. Thanks again!
